src/simsopt/field/force.py

```diff
@@ -2,55 +2,182 @@
 from scipy import constants
 import numpy as np
 import jax.numpy as jnp
-from jax import grad
+import jax.scipy as jscp
+from jax import grad, vmap
+from jax.lax import cond
 from .biotsavart import BiotSavart
 from .selffield import B_regularized_pure, B_regularized, regularization_circ, regularization_rect
 from ..geo.jit import jit
 from .._core.optimizable import Optimizable
 from .._core.derivative import derivative_dec
-
 Biot_savart_prefactor = constants.mu_0 / 4 / np.pi
 
+__all__ = [
+    "coil_force",
+    "self_force_circ",
+    "self_force_rect",
+    "coil_coil_inductances_pure",
+    "coil_coil_inductances_inv_pure",
+    "induced_currents_pure",
+    "NetFluxes",
+    "B2_Energy",
+    "coil_net_force",
+    "coil_net_torque",
+    "coil_torque",
+    "LpCurveForce_deprecated",
+    "MeanSquaredForce_deprecated",
+    "SquaredMeanForce",
+    "LpCurveForce",
+    "SquaredMeanTorque",
+    "LpCurveTorque",
+]
+
+
+def coil_force(coil, other_coils):
+    """
+    Compute the force on a coil.
+
+    Args:
+        coil (Coil): Coil to compute the pointwise forces on.
+        other_coils (list): List of coils contributing forces on the primary coil.
+        regularization (Regularization): Regularization object.
 
-def coil_force(coil, allcoils, regularization):
+    Returns:
+        array: Array of force.
+    """
     gammadash = coil.curve.gammadash()
     gammadash_norm = np.linalg.norm(gammadash, axis=1)[:, None]
     tangent = gammadash / gammadash_norm
-    mutual_coils = [c for c in allcoils if c is not coil]
-    mutual_field = BiotSavart(mutual_coils).set_points(coil.curve.gamma()).B()
-    mutualforce = np.cross(coil.current.get_value() * tangent, mutual_field)
-    selfforce = self_force(coil, regularization)
-    return selfforce + mutualforce
+    mutual_coils = [c for c in other_coils if c is not coil]
+
+    ### Line below seems to be the issue -- all these BiotSavart objects seem to stick
+    ### around and not to go out of scope after these calls!
+    mutual_field = BiotSavart(mutual_coils).set_points(coil.curve.gamma())
+    B_mutual = mutual_field.B()
+    mutualforce = np.cross(coil.current.get_value() * tangent, B_mutual)
+    selfforce = self_force(coil)
+    return (selfforce + mutualforce)
+
+
+def coil_net_force(coil, other_coils):
+    """
+    Compute the net forces on a list of coils.
```

Comment:
can we include units? can we clarify that this is the integrated pointwise force on a coil?